src/architecture/merkledb.md

```diff
@@ -0,0 +1,207 @@
+# Exonum MerkleDB
+
+Exonum MerkleDB is a persistent storage implementation based on
+[RocksDB][rocks-db]. MerkleDB provides APIs to work with merkelized data
+structures.
+
+MerkleDB is an object database. Objects represent the highest abstraction level
+for the [data storage](storage.md). All objects fall into the following groups:
+
+- **blobs** which represent sequences of bytes
+- **root objects** that do not have parents. These objects have UTF-8
+  identifiers, for example, "block", "state". Root objects can contain blob
+  items inside them.
+
+## Root Object Types
+
+Root objects perform the same role as in relational database
+management systems (RDBMSs). However, unlike RDBMS tables,
+all MerkleDB objects internally are implemented as wrappers around key-value
+stores.
+
+Root objects in MerkleDB fall into several types:
+
+- **list**, where either keys or values are stored
+- **map**, where key-value pairs are stored
+- **set** of unique items
+- **entry**, which represents a set containing only one item.
+
+Both keys and values in the wrapped stores are persisted as byte sequences.
+MerkleDB does not natively support operations (matching, grouping, sorting,
+etc.) over separate value fields, as it is the case with other key-value
+storages.
+
+### Key Sorting and Iterators
+
+MerkleDB objects implement iterators over stored items. The stored items are:
+
+- keys or values in case of list objects, sets or entries
+- key-value pairs in case of map objects.
+
+Such iterators use key ordering of the underlying key-value storage to
+determine the iteration order. Namely, keys are lexicographically ordered
+according to their binary serializations. This ordering coincides with that
+used in [RocksDB](storage.md).
+
+### MapIndex
+
+[`MapIndex`][map-index] implements a key-value store aka a map. It has
+the following functionality:
+
+- Get, set and remove value by key
+- Check if a specific key is present in the map
+- Iterate over the key-value pairs in the lexicographic key order
+- Iterate over keys in the lexicographic key order
```

Comment:
What if we join these together:

"Iterate over keys, values and entries in …"

Reply:
> Also now there is possible to get total number of items for every collection type.

done